manpages/freesitemgr.1
.TH "freesitemgr" "1" "0.1.4" "Freenet Project" "pyFreenet"
.SH "NAME"
.LP 
freesitemgr \- Insert or re\-insert freesites into freenet

.SH "SYNOPSIS"
.LP 
\fBfreesitemgr\fP [\fIoptions\fP] \fIcommand\fP [\fIarguments\fP]
.SH "DESCRIPTION"
.LP 
\fBfreesitemgr\fP is a simple command\-line freenet client program
for inserting \fBfreesites\fP.

Use of this program is controlled by the \fIcommand\fP argument
you give it.

run \fBfreesitemgr \-h\fP for a detailed help output.

Note that this is a new, rewritten version of freesitemgr, designed
to be extremely fault\-tolerant. If you interrupt freesitemgr,
or if the freenet node dies, halfway through an insert, then you can
simply re\-run freesitemgr, and it will pick up where it left off.
This new resilience is a response to public demand, and the needs
of users wanting to insert large freesites with high file counts.

.SH "OPTIONS"
.LP 
.TP 
\fB\-h\fR, \fB\-\-help\fR
Print help information and exit
.TP 
\fB\-V\fR, \fB\-\-version\fR
Print version number and exit
.TP 
\fB\-c\fR, \fB\-\-config\-dir=\fIdir\fR
Sets config directory (default ~/.freesitemgr/)
.TP 
\fB\-v\fR, \fB\-\-verbose\fR
Enable verbose output to the stderr stream. Can be given
twice to enable FCP transcripts, or more to enable more
detailed debugging information
.TP 
\fB\-q\fR, \fB\-\-quiet\fR
Run quietly
.TP 
\fB\-r\fR, \fB\-\-priority=\fIpriority\fR
Set the FCP priority for inserts. Valid values are between
0 (highest) to 6 (lowest), default is 3. For large freesites,
please do not set this to 0, 1 or 2.
.TP 
\fB\-l\fR, \fB\-\-logfile=\fIlogfile\fR
location of logfile (default stdout)
.TP 
\fB\-C\fR, \fB\-\-cron\fR
Set options suitable for use of freesitemgr in your crontab.
This will cause freesitemgr to output dated headers and set
one verbosity level.

For example, to run freesitemgr at 4am each day,
you could put in your crontab:

\fB0 4 * * * freesitemgr \-C update >> ~/freesitemgr.log\fR
.TP 
\fB\-\-chk\-calculation\-node=\fIhostname[:port]\fR
Sets a different freenet node to be used for CHK pre\-calculation phase of
site insertion. This can save huge amounts of time
in cases where one is inserting to
a remote node over a relatively slow link, but has access to a closer node
on a faster link (or running on the local machine), and doesn't want
to use the closer node for the actual inserts.

.LP 

.SH "COMMANDS"
.LP 
.TP 

\fBhelp\fP
Synonym for \fB\-h\fR or \fB\-\-help\fR, displays a help page
.TP 

\fBsetup\fP
Create or modify the freesite config file (\fB~/.freesitemgr/.config\fP)
interactively.
.TP 

\fBadd\fP
Add a new freesite interactively
.TP 

\fBlist \fIsitename [sitename...]\fP
display a summary of all freesites, or a
detailed report of one or more site(s) if \fIsitename\fR given
.TP 

\fBlistall\fP
display a detailed report of all freesites
.TP 

\fBremove \fIsitename\fP [\fIsitename\fR...]
remove given freesite
.TP 

\fBupdate \fIsitename\fP [\fIsitename\fR...]
This command checks on any existing
update jobs, and ticks required items off the queue. Also, it
checks all freesites for changes, and if any have changed, it
launches insert jobs that are required to bring the freesite up to date.
If the freesite was already being inserted, no action will be taken.
Note that this command will not wait for new/changed freesites to
be fully inserted, it will instead exit and leave the insertions
on the freenet node global queue.
You can safely add a 'freesitemgr update' command to your crontab.
If one or more \fIsitename\fR values are given, then only those
freesites will be updated. Otherwise, all freesites will be updated.
.TP 

\fBcancel \fIsitename\fR [\fIsitename\fR...]
Cancel the insertion of one or more
site(s) \fIsitename\fR. Removes its entries
from the freenet node global queue, and reset its control file
back to an idle state.
.TP 

\fBcleanup \fIsitename\fR [\fIsitename\fR...]
Cleans up the node queue in respect of inserts for the given site(s)
\fIsitename\fR. Removes queue entries for inserts that have been
completed, and updates local site record in ~/.freesitemgr/\fIsitename\fR.

.LP 

.SH "FILES"
freesitemgr keeps track of all freesite insertions in a directory
chosen by the user.

The default directory is \fB~/.freesitemgr/\fP. All files within this
directory are in Python syntax (for easy reading and hacking).

This directory contains a file called \fB.config\fP, which holds
global configuration information for freesitemgr.

Each other file has a name which is the name of the freesite as allocated
by the user. For instance, a freesite \fBfoo\fR will be tracked 
in the file \fB~/.freesitemgr/foo\fR.

.SH "ENVIRONMENT VARIABLES"
.LP 
.TP 
\fBFCP_HOST\fP
Specifies the hostname of the FCP interface. This value
will be used unless overridden with '\-H'.
.TP 
\fBFCP_PORT\fP
Specifies the port number of the FCP interface. This value
will be used unless overridden with '\-P'.

.LP 
.SH "AUTHORS"
.LP
The pyFreenet applications and library were originally written
by David McNab (aum) in May-June, 2006 and are maintained
by the Freenet Project team.
.LP
Contact the original author at: <davidmcnabnz@gmail.com>
.SH "REPORTING BUGS"
.LP
If you find a bug, please post it to the Freenet bug tracker at:
.RS
.LP    
https://bugs.freenetproject.org/bug_report_page.php
.RE
.LP
but first please be sure to target the 'pyFreenet' project, by
selecting 'pyFreenet' in the combobox on the top-right of the page.
.SH "SEE ALSO"
.LP 
fcpget(1) fcpput(1) fcpgenkey(1) fcpinvertkey(1) fcpredirect(1)



CHANGELOG

Revision history for pyFreenet

- Version 0.2.5
    - added 'noderefbot' - an IRC bot which swaps noderefs automatically

- Version 0.2.4

    - minor bugfix release
    - freesitemgr: now saving updated freesite edition numbers in
      ~/.freesitemgr/<sitename>

- Version 0.2.3

    - added 'fproxyproxy' shim atop fproxy, which should be used as
      a browser proxy, not surfed directly

- Version 0.2.2

    - added 'name services layer' (see 'fcpnames' manpage)
    - integrated this layer into .get()/.put() methods (and
      consequently, into fcpget and fcpput
    - added the utility 'fcpnames', allowing easy management
      of name services

- Version 0.2.1

    - added commands 'help' and 'cleanup' to freesitemgr
    - moved fcpget/fcpput/etc into modules in package fcp, leaving simple
      2-liner front-ends

- Version 0.2

    - added another layer of recovery to freesitemgr, to re-submit
      jobs the node forgets about (ie, losts from its queue)
    - added '-C' option to freesitemgr, to suit its use in crontab entries

- Version 0.1.8

    - another rework of freesitemgr:
        - inserts use only the global queue
        - 'freesitemgr update' exits immediately, leaving jobs on queue
        - extensive accounting, which recognises and compensates for node
          failures
        - you can now safely cron 'freesitemgr update'

- Version 0.1.7

    - rewrote most of freesitemgr:
        - now uses directory ~/.freesitemgr/
        - keeps track of all inserts
        - if interrupted, or if node dies, will resume where
          it left off
        - simplified options

- Version 0.1.6

    - now storing manifests in .freesites (for freesitemgr)
    - freesitemgr now issues progress messages with '-v' set
    - all utils now accept multiple occurrences of '-v' option, with
      increasing verbosity
    - added timeout option to get/put (and '-t' to fcpget/fcpput)

- Version 0.1.5

    - added global queue and persistence support for fcpget/fcpput
    - added 'nowait' option to fcpput

- Version 0.1.4

    - added manpages for console programs
    - improved mimetype determination algorithm for fcp put
    - added 'freedisk', a rudimentary linux filesystem that maps freenet
      into a mountable fs - limited functionality at present
    - added support for use of env vars FCP_HOST and FCP_PORT for specifying
      FCP host/port - useful for people who access FCP across a LAN, since
      it avoids annoyance of having to specify -H or -P with each command

- Version 0.1.3

    - added 'fcpget' and 'fcpput' command-line key retrieve/insert apps
    - added 'freesitemgr' command-line freesite insertion app
    - several bug fixes

- Version 0.1.2

    - added xmlrpc server app
    - added xmlrpc server CGI module (for embedding a Freenet XML-RPC
      server into websites)
    - added 'freesitemgr', a console freesite insertion app

- Version 0.1.1
    - 2006-May-13
        - First packaged release



freesitemgr.py
#!/usr/bin/env python
#@+leo-ver=4
#@+node:@file freesitemgr.py
#@@first
#@+others
#@+node:freesitemgr-script
"""
A utility to update freesites from within a cron environment
"""
#@+others
#@+node:imports
import sys, os, time, commands, traceback, getopt

import fcp.node
from fcp.sitemgr import SiteMgr, fixUri

#@-node:imports
#@+node:globals
progname = sys.argv[0]

# time we wait after starting fred, to allow the node to 'warm up'
# and make connections to its peers
startupTime = 180

# directory where we have freenet installed,
# change it as needed
#freenetDir = "/home/david/freenet"

homeDir = os.path.expanduser("~")

# derive path of freenet pid file, the (non)existence
# of which is the easiest test of whether the freenet
# node is running
#freenetPidFile = os.path.join(freenetDir, "Freenet.pid")

logFile = os.path.join(homeDir, "updatesites.log")
pidFile = os.path.join(homeDir, "updatesites.pid")

if sys.platform.startswith("win"):
    confDirName = "freesitemgr"
else:
    confDirName = ".freesitemgr"
confDir = os.path.join(homeDir, confDirName)

#@-node:globals
#@+node:editCreateConfig
def editCreateConfig(sitemgr):
    """
    Creates an initial config file interactively
    """
    print "Setting up configuration file %s" % sitemgr.conffile

    # get fcp hostname
    fcpHost = raw_input("FCP Hostname [%s] (* for all): " % sitemgr.fcpHost).strip()
    if not fcpHost:
        fcpHost = sitemgr.fcpHost
    if fcpHost == '*':
        fcpHost = ""
    
    # get fcp port
    while 1:
        fcpPort = raw_input("FCP Port [%s]: " % sitemgr.fcpPort).strip()
        if not fcpPort:
            fcpPort = sitemgr.fcpPort
        try:
            fcpPort = int(fcpPort)
        except:
            continue
        break

    print "Trying FCP port at %s:%s" % (fcpHost, fcpPort)
    try:
        fcpnode = fcp.FCPNode(host=fcpHost, port=fcpPort)
    except:
        traceback.print_exc()
        print "Failed to connect to FCP Port"
        print "Please ensure your node is running, with its FCP port"
        print "reachable at %s:%s, and try this command again" % (fcpHost, fcpPort)
        print "Setup aborted"
        return

    fcpnode.shutdown()

    sitemgr.fcpHost = fcpHost
    sitemgr.fcpPort = fcpPort

    # confirm and save
    if getyesno("Save configuration", True):
        sitemgr.save()

    print "Configuration saved to %s" % sitemgr.conffile

#@-node:editCreateConfig
#@+node:addSite
def addSite(sitemgr):
    """
    Interactively adds a new site to config
    """
    print "Add new site"
    
    if not sitemgr.node:
        print "Cannot add site - no contact with node on %s:%s" % (
            sitemgr.fcpHost, sitemgr.fcpPort)
        print "Please ensure your freenet node is running, or run"
        print "'%s setup' to edit your FCP access address" % progname

    while 1:
        sitename = raw_input("Name of freesite, or empty line to cancel: ").strip()
        if not sitename:
            print "Add site aborted"
            return
        elif sitemgr.hasSite(sitename):
            print "Freesite '%s' already exists" % sitename
            continue
        break

    while 1:
        sitedir = raw_input("Directory where freesite's files reside: ").strip()
        if not sitedir:
            print "Add site aborted"
            return
        sitedir = os.path.abspath(sitedir)
        if not os.path.isdir(sitedir):
            print "'%s' is not a directory, try again" % sitedir
            continue
        #elif not os.path.isfile(os.path.join(sitedir, "index.html")):
        #    print "'%s' has no index.html, try again" % sitedir
        #    continue
        break

    while 1:
        uriPriv = raw_input("Site private URI (if any - press ENTER for new\n: ").strip()
        if not uriPriv:
            uriPub, uriPriv = sitemgr.node.genkey()
        else:
            if not fcp.node.uriIsPrivate(uriPriv):
                print "Sorry, that's a public URI, we need a private URI"
                continue
            try:
                uriPub = sitemgr.node.invertprivate(uriPriv)
            except:
                traceback.print_exc()
                print "Invalid private URI:\n  %s" % uriPriv
                continue
        break
        uriPub = fixUri(uriPub, sitename)
        uriPriv = fixUri(uriPriv, sitename)
    
    # good to go - add the site
    sitemgr.addSite(name=sitename, dir=sitedir, uriPub=uriPub, uriPriv=uriPriv)

    print "Added new freesite: '%s' => %s" % (sitename, sitedir)

#@-node:addSite
#@+node:removeSite
def removeSite(sitemgr, sitename):
    """
    tries to remove site from config
    """
    if not sitemgr.hasSite(sitename):
        print "No such freesite '%s'" % sitename
        return

    if getyesno("Are you sure you wish to delete freesite '%s'" % sitename, False):
        sitemgr.removeSite(sitename)
        print "Removed freesite '%s'" % sitename

#@-node:removeSite
#@+node:cancelUpdate
def cancelUpdate(sitemgr, sitename, force=False):
    """
    tries to remove site from config
    """
    if not sitemgr.hasSite(sitename):
        print "No such freesite '%s'" % sitename
        return

    doit = False
    if force:
        doit = True
    elif getyesno("Are you sure you wish to cancel update for freesite '%s'" \
                    % sitename, False):
        doit = True
    if doit:
        sitemgr.cancelUpdate(sitename)
        print "Cancelled update for freesite '%s'" % sitename
    else:
        print "Not cancelling update for freesite '%s'" % sitename

#@-node:cancelUpdate
#@+node:getChkCalcNode
def getChkCalcNode(addr):
    """
    yuck - using a separate node for chk calculation
    """
    parts = addr.split(":")
    nparts = len(parts)
    if nparts == 1:
        chkHost = addr
        chkPort = fcp.node.defaultFCPPort
    elif nparts == 2:
        chkHost = parts[0] or fcp.node.defaultFCPHost
        chkPort = parts[1] or fcp.node.defaultFCPPort
    
    try:
        chkNode = fcp.node.FCPNode(host=chkHost, port=chkPort)
        return chkNode
    except:
        return None

#@-node:getChkCalcNode
#@+node:getYesNo
def getyesno(ques, default=False):
    """
    prompt for yes/no answer, with default
    """
    if default:
        prmt = "[Y/n]"
    else:
        prmt = "[y/N]"
        
    resp = raw_input(ques + " " + prmt + " ").strip().lower()
    
    if not resp:
        return default
    #elif resp[0] in ['y', 't']:
    elif resp[0] == 'y':
        return True
    else:
        return False
#@-node:getYesNo
#@+node:help
def help():
    """
    dump help info and exit
    """
    print "%s: a console-based USK freesite insertion utility" % progname
    
    print "Usage: %s [options] <command> <args>" % progname
    print "Options:"
    print "  -h, --help"
    print "          - display this help message"
    print "  -c, --config-dir=path"
    print "          - use different config directory (default is %s)" % confDir
    print "  -v, --verbose"
    print "          - run verbosely, set this twice for even more noise"
    print "  -q, --quiet"
    print "          - run quietly"
    print "  -l, --logfile=filename"
    print "          - location of logfile (default %s)" % logFile
    print "  -r, --priority"
    print "     Set the priority (0 highest, 6 lowest, default 4)"
    print "     of 'forever', so the insert will resume if the node crashes"
    print "  -C, --cron"
    print "     Set options suitable for putting freesitemgr in your crontab,"
    print "     and output a dated header with each site insert"
    print "  --chk-calculation-node=hostname[:port]"
    print "     Use a different node for CHK calculations, which can be a"
    print "     timesaver when inserting large amounts of data into a remote node"
    print
    print "Available Commands:"
    print "  setup              - create/edit freesite config file interactively"
    print "  add                - add new freesite - user will be prompted for"
    print "                       details"
    print "  list [<name>]      - display a summary of all freesites, or a"
    print "                       detailed report of one site if <name> given"
    print "  listall            - print detailed report of all sites"
    print "  remove <name>      - remove given freesite"
    print "  update [<name>...] - reinsert freesites which have changed since"
    print "                       they were last inserted. If no site names are"
    print "                       given, then all freesites will be updated"
    print "  cancel <name>...   - cancel any pending insert of freesite(s) <name>."
    print "  cleanup <name>...  - clean up node queue for site(s) <name>"
    print "  help               - same as '-h', display this help page"

    sys.exit(0)

#@-node:help
#@+node:usage
def usage(ret=-1, msg=None):
    if msg != None:
        print msg
    print "Usage: %s [options] <command> [<arguments>]" % progname
    print "Do '%s -h' for help" % progname

    sys.exit(ret)

#@-node:usage
#@+node:noNodeError
def noNodeError(sitemgr, msg):
    print msg + ": cannot connect to node FCP port"
    print "To use this command, you must have a freenet node running"
    print "Please ensure your node FCP port is reachable at %s:%s" % (
        sitemgr.fcpHost, sitemgr.fcpPort)
    print "or run '%s setup' to configure a different FCP port" % progname
    sys.exit(1)

#@-node:noNodeError
#@+node:main
def main():

    force = False
    cron = False
    chkCalcNode = None

    # default job options
    opts = {
            "verbosity" : fcp.node.ERROR,
            "Verbosity" : 0,
            #"logfile" : logFile,
            'priority' : 3,
            }

    # process command line switches
    try:
        cmdopts, args = getopt.getopt(
            sys.argv[1:],
            "?hvc:l:r:fCV",
            ["help", "verbose", "config-dir=", "logfile=",
             "max-concurrent=", "force",
             "priority", "cron",
             "chk-calculation-node=",
             "version",
             ]
            )
    except getopt.GetoptError:
        # print help information and exit:
        usage()
        sys.exit(2)
    output = None
    verbose = False
    #print cmdopts
    for o, a in cmdopts:

        if o in ("-?", "-h", "--help"):
            help()
            sys.exit(0)

        if o in ("-V", "--version"):
            print "This is %s, version %s" % (progname, fcp.node.fcpVersion)
            sys.exit(0)

        if o in ("-v", "--verbosity"):
            opts['verbosity'] += 1
            opts['Verbosity'] = 1023
        
        if o in ("-q", "--quiet"):
            opts['verbosity'] = fcp.node.SILENT
        
        if o in ("-c", "--config-dir"):
            opts['basedir'] = a
        
        if o in ("-l", "--logfile"):
            opts['logfile'] = a

        if o == '--chk-calculation-node':
            chkNode = getChkCalcNode(a)
            if not chkNode:
                usage("Failed to connect to specified CHK calc node '%s'" % a)
            opts['chkCalcNode'] = chkNode
        
        if o in ("-r", "--priority"):
            try:
                pri = int(a)
                if pri < 0 or pri > 6:
                    raise hell
            except:
                usage("Invalid priority '%s'" % pri)
            opts['priority'] = int(a)

        if o in ("-f", "--force"):
            force = True

        if o in ("-C", "--cron"):
            opts['verbosity'] = fcp.node.INFO
            opts['Verbosity'] = 1023
            cron = True

    # process command
    if len(args) < 1:
        usage(msg="No command given")

    cmd = args.pop(0)

    if cmd not in [
            'setup','config','init',
            'add',
            'remove',
            'list', 'listall',
            'update',
            'cancel', "help", "cleanup",
            ]:    
        usage(msg="Unrecognised command '%s'" % cmd)

    # we now have a likely valid command, so now we need a sitemgr
    sitemgr = SiteMgr(**opts)

    if cmd in ['setup', 'init', 'config']:
        editCreateConfig(sitemgr)

    elif cmd == 'help':
        help()
        sys.exit(0)

    elif cmd == 'add':
        if not sitemgr.node:
            noNodeError(sitemgr, "Cannot add site")
        addSite(sitemgr)

    elif cmd == 'remove':
        if not args:
            print "Remove site: no freesites selected"
            return
        for sitename in args:
            removeSite(sitemgr, sitename)

    elif cmd == 'cancel':
        if not args:
            print "Cancel site update: no freesites selected"
            return
        for sitename in args:
            cancelUpdate(sitemgr, sitename, force)

    elif cmd == 'cleanup':
        sitemgr.cleanup(*args)

    elif cmd in ['list', 'listall']:
        if cmd == 'listall':
            sites = [site.name for site in sitemgr.sites]
        else:
            sites = args
        if not sites:
            # summary list
            names = []
            for site in sitemgr.sites:
                if site.updateInProgress:
                    names.append("*"+site.name)
                else:
                    names.append(site.name)
            print " ".join(names)
        else:
            for sitename in sites:
                if not sitemgr.hasSite(sitename):
                    print "No such site '%s'" % sitename
                else:
                    site = sitemgr.getSite(sitename)
                    if site.updateInProgress:
                        state = "updating"
                    else:
                        state = "idle"
                    print "%s:" % sitename
                    print "    state: %s" % state
                    print "    dir: %s" % site.dir
                    print "    uri: %s" % site.uriPub
                    print "    privkey: %s" % site.uriPriv
                    #print "    version: %s" % info['version']

            pass

    elif cmd == 'update':
        if not sitemgr.node:
            noNodeError(sitemgr, "Cannot update freesites")
        try:
            if not args:
                sites = sitemgr.getSiteNames()
            else:
                sites = args
            sitemgr.insert(cron=cron, *args)
        except KeyboardInterrupt:
            print "freesitemgr: site inserts cancelled by user"

    try:
        sitemgr.node.shutdown()
    except:
        traceback.print_exc()
        try:
            sitemgr.node.socket.close()
        except:
            pass
        pass

#@-node:main
#@+node:mainline
if __name__ == '__main__':
    main()

#@-node:mainline
#@-others
#@-node:freesitemgr-script
#@-others
#@-node:@file freesitemgr.py
#@-leo
